### PR TITLE
allow custom mutations through actions

### PR DESCRIPTION
**Pull request by 0x777**
**Needs https://github.com/hasura/graphql-parser-hs/pull/19 to be merged**

I'll be pushing changes to this PR over the next couple of days. A high level overview of actions can be found in this PR's [docs](https://deploy-preview-3042--hasura-docs.netlify.com/graphql/manual/actions/index.html). 

### Updated: 23/10/2019
Tasks done:
- [x] defining custom types
- [x] sync actions
- [x] async actions
- [x] metadata export/import
- [x] http endpoint as action handlers
- [x] relationships

Remaining:
- [ ] pg functions as action handlers
- [ ] async action relationships

`update_action`:-
```yaml
type: update_action
args:
  name: action-name
  definition: action-definition # will replaces existing one
```

The webhook is called over `POST`. The payload looks like:-
```json
{
  "session_variables": {
    "x-hasura-user-id": "1234"
  },
  "input": {
    "arg1": "value1",
    "arg2": "value2"
  },
  "action": {
    "name": "action-name"
  }
}
```

**The webhook response:**
- On success, should return a `2xx` code with JSON body whose value is of the defined output type
- Should return a `4xx` code for an error that should be propagated back to the client. The response JSON body is an object with a compulsory `message` field (text) and optional `code` field (text)
- All other status codes are treated as internal errors and nothing is propagated to the client. The details should be logged as `internal` in `error` field and part of extensions in response for the admin role. 

**Single row mutations:-** Implemented the RFC https://github.com/hasura/graphql-engine/issues/3731

### CODE REVIEW

**Comment by 0x777** [comment]
https://deploy-preview-3042--hasura-docs.netlify.com/graphql/manual/actions/index.html

**Comment by wawhal** [comment]
Hey @0x777 I think I found a bug.

If my output type of any action is `UserInfo`, the server will not allow me to remove the type `UserInfo` from custom types. But if the output type is `UserInfo!`, the server lets me remove `UserInfo` from custom types.

Expected behaviour is that if any extension of `UserInfo` is used by an action, the server should not allow me to remove them from custom types.

File: server/src-lib/Hasura/GraphQL/Resolve/Action.hs [hunk 1]
```diff
@@ -315,23 +319,25 @@ callWebhook manager reqHeaders confHeaders forwardClientHeaders resolvedWebhook
...
+
+      if | HTTP.statusIsSuccessful responseStatus  -> do
+             maybeData <- _awsData <$> decodeValue responseValue
+             onNothing maybeData $ throw500WithDetail "internal error" $
+               "webhook response does not have 'data' key for 2xx response status"
+
+         | HTTP.statusIsClientError responseStatus -> do
+             maybeError <- _aweErrors <$> decodeValue responseValue
+             ActionWebhookErrorResponse message maybeCode <-
+               onNothing maybeError $ throw500WithDetail "internal error" $
```
Comment (tirumaraiselvan) [inline server/src-lib/Hasura/GraphQL/Resolve/Action.hs]: shouldn't this be "unknown error" and not "internal error" with 4xx response code?

File: server/src-lib/Hasura/GraphQL/Resolve/Action.hs [hunk 2]
```diff
@@ -315,23 +319,25 @@ callWebhook manager reqHeaders confHeaders forwardClientHeaders resolvedWebhook
...
-              withMessage message = err500 code message
-              noMessagekey = "\"message\" key is not found in webhook \"errors\" response"
-              withoutMessage = (err500 code noMessagekey)
-                               {qeInternal = Just $ J.object ["webhook_response" J..= responseValue]}
-          throwError $ maybe withoutMessage withMessage maybeMessage
+          responseStatus = responseWreq ^. Wreq.responseStatus
+
+      if | HTTP.statusIsSuccessful responseStatus  -> do
+             maybeData <- _awsData <$> decodeValue responseValue
+             onNothing maybeData $ throw500WithDetail "internal error" $
```
Comment (tirumaraiselvan) [inline server/src-lib/Hasura/GraphQL/Resolve/Action.hs]: this is going to be an interesting case where webhook returns 200 but hasura returns 500.

Comment (tirumaraiselvan) [reply server/src-lib/Hasura/GraphQL/Resolve/Action.hs]: Discussed with Vamshi, anything that breaks response contract is 500 for the client.

**Comment by followbl** [comment]
Pretty excited to try this out...think this has the potential to clean-up a lot of code for us that *feels* like it's in the wrong place right now

**Comment by mrjjwright** [comment]
This looks so awesome.  Can really build little empires with this approach.

File: docs/graphql/manual/actions/async-actions.rst [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+
+
+.. contents:: Table of contents
+  :backlinks: none
+  :depth: 1
+  :local:
+
+WORK IN PROGRESS
+
+Sometimes you may not want to wait for an action to complete (say if the business logic takes a long time). In such cases you can create an **asynchronous** action, which returns an ``action_id`` immediately to the client before contacting the webhook.
```
Comment (tsaiDavid) [inline docs/graphql/manual/actions/async-actions.rst]: Love this feature! In trying this, I noticed that when running an async mutation, I am seeing `id` being made available through the GraphQL schema when attempting to subscribe to my action - as opposed to the `action_id` stated here.

**Comment by weyert** [comment]
Thank you, this feature looks really interesting. If I want to experiment with this what's the best way? Can I always just use the latest `hasura/graphql-engine:pull3042-be07e042`? Do you have any estimation when you want to have this released? Would it be part of v1.1?

**Comment by jascodes** [comment]
@weyert Try with https://github.com/hasura/graphql-engine/pull/3252

**Comment by bgever** [comment]
Nice work folks! Looking forward to give this a spin. First thought that came to mind is **how to apply custom authorization rules based on a JWT** that's sent along with the request. I see there's a placeholder in the [Sample use cases](https://deploy-preview-3042--hasura-docs.netlify.com/graphql/manual/actions/use-cases.html) WIP docs, hope that can be populated soon.

Keep up the amazing work! 👍

**Comment by wawhal** [comment]
@bgever you get all the session variables from the JWT as a part of the request payload to your webhook. The payload looks like:

```json
{
  "session_variables": {
    "x-hasura-user-id": "1234",
    "x-hasura-role": "user"
  },
  "input": {
    "arg1": "value1",
    "arg2": "value2"
  },
  "action": {
    "name": "action-name"
  }
}
```

File: server/src-lib/Hasura/GraphQL/Resolve/Mutation.hs [hunk 4]
```diff
@@ -204,17 +209,49 @@ convertUpdate opCtx fld = do
...
      , Has OrdByCtx r, Has SQLGenCtx r
      )
+  => UpdOpCtx -- the update context
+  -> Field -- the mutation field
+  -> m RespTx
+convertUpdateByPk opCtx field = do
+  responseTx <- convertUpdateGeneric opCtx boolExpParser tableSelectionAsMutationFields field
+  pure $ do
+    response <- responseTx
+    RM.withSingleTableRow response
```
Comment (rakeshkky) [inline server/src-lib/Hasura/GraphQL/Resolve/Mutation.hs]: Improve `withSingleTableRow`.

File: server/src-lib/Hasura/RQL/DML/Select/Types.hs [hunk 5]
```diff
@@ -105,7 +105,7 @@ data ComputedFieldScalarSel v
 
 data ComputedFieldSel v
   = CFSScalar !(ComputedFieldScalarSel v)
-  | CFSTable !(AnnSimpleSelG v)
+  | CFSTable !Bool !(AnnSimpleSelG v)
```
Comment (rakeshkky) [inline server/src-lib/Hasura/RQL/DML/Select/Types.hs]: Use a sum type instead of `Bool`

File: server/src-lib/Hasura/GraphQL/Resolve/Types.hs [hunk 6]
```diff
@@ -218,13 +234,41 @@ data UnresolvedVal
...
-$(makePrisms ''ResolveField)
-$(makeLenses ''ComputedField)
-$(makePrisms ''ComputedFieldType)
-
 data InputFunctionArgument
   = IFAKnown !FunctionArgName !UnresolvedVal -- ^ Known value
   | IFAUnknown !FunctionArgItem -- ^ Unknown value, need to be parsed
   deriving (Show, Eq)
+
+-- instance Semigroup QueryReusability where
```
Comment (rakeshkky) [inline server/src-lib/Hasura/GraphQL/Resolve/Types.hs]: remove commented code.

File: server/src-lib/Hasura/GraphQL/Schema.hs [hunk 7]
```diff
@@ -691,21 +719,27 @@ noFilter = annBoolExpTrue
...
+  => AnnotatedObjects -> TableCache -> FunctionCache -> ActionCache -> m GCtxMap
+mkGCtxMap annotatedObjects tableCache functionCache actionCache = do
   typesMapL <- mapM (mkGCtxMapTable tableCache functionCache) $
                filter (tableFltr . _tiCoreInfo) $ Map.elems tableCache
-  typesMap <- combineTypes typesMapL
-  return $ flip Map.map typesMap $ \(ty, flds, insCtxMap) ->
-    mkGCtx ty flds insCtxMap
+  actionsSchema <- mkActionsSchema annotatedObjects actionCache
+  typesMap <- combineTypes actionsSchema typesMapL
+  -- TODO: clean this up
```
Comment (rakeshkky) [inline server/src-lib/Hasura/GraphQL/Schema.hs]: Move this comment to right place.

File: server/src-lib/Hasura/GraphQL/Schema/Action.hs [hunk 8]
```diff
@@ -0,0 +1,284 @@
...
+  ( actionExecutionContext
+  , fieldInfo
+  )
+  where
+    definition = _aiDefinition actionInfo
+    actionExecutionContext =
+      case _adKind definition of
+        ActionSynchronous  ->
+          ActionExecutionSyncWebhook $ SyncActionExecutionContext actionName
+  -- TODO: only covers object types
```
Comment (rakeshkky) [inline server/src-lib/Hasura/GraphQL/Schema/Action.hs]: Not required.

File: server/src-lib/Hasura/GraphQL/Schema/Action.hs [hunk 9]
```diff
@@ -0,0 +1,284 @@
...
+          definitionList
+          (_adHandler definition)
+          (_adHeaders definition)
+          (_adForwardClientHeaders definition)
+        ActionAsynchronous -> ActionExecutionAsync $ _apiFilter permission
+
+    -- TODO: we need to capture the comment from action definition
```
Comment (rakeshkky) [inline server/src-lib/Hasura/GraphQL/Schema/Action.hs]: Do this.

File: server/src-lib/Hasura/GraphQL/Schema/Action.hs [hunk 10]
```diff
@@ -0,0 +1,284 @@
...
+           , mkHsraObjFldInfo (Just description) (unActionName actionName)
+             (mapFromL _iviName [idArgument])
+             (G.toGT $ G.toGT $ mkAsyncActionSelectionType actionName)
+
+           , TIObj $ mkAsyncActionQueryResponseObj actionName $
+             _adOutputType definition
+           )
+    ActionSynchronous -> Nothing
+  where
+    -- TODO: we need to capture the comment from action definition
```
Comment (rakeshkky) [inline server/src-lib/Hasura/GraphQL/Schema/Action.hs]: Do this.

File: server/src-lib/Hasura/GraphQL/Schema/Action.hs [hunk 11]
```diff
@@ -0,0 +1,284 @@
...
+  -> m (Map.HashMap RoleName
+         ( Maybe (ActionSelectOpContext, ObjFldInfo, TypeInfo)
+         , (ActionExecutionContext, ObjFldInfo)
+         , FieldMap
+         )
+       )
+mkActionSchemaOne annotatedObjects actionInfo = do
+  -- annotatedOutputTypeM <- case _aiOutputTypeInfo actionInfo of
+  --   ActionOutputObject _ ->
+  -- annotatedOutputTypeM <- fmap Just $ onNothing
```
Comment (rakeshkky) [inline server/src-lib/Hasura/GraphQL/Schema/Action.hs]: remove comments

File: server/src-lib/Hasura/GraphQL/Schema/Action.hs [hunk 12]
```diff
@@ -0,0 +1,284 @@
...
+  annotatedOutputType <- onNothing
+      (Map.lookup (ObjectTypeName actionOutputBaseType) annotatedObjects) $
+      throw500 $ "missing annotated type for: " <> showNamedTy actionOutputBaseType
+    -- _ -> return Nothing
```
Comment (rakeshkky) [inline server/src-lib/Hasura/GraphQL/Schema/Action.hs]: remove comment

File: server/src-lib/Hasura/GraphQL/Validate/Types.hs [hunk 13]
```diff
@@ -609,14 +632,13 @@ mkTyInfoMap tyInfos =
...
 fromTyDef :: G.TypeDefinition -> TypeLoc -> Either Text TypeInfo
-fromTyDef tyDef loc = case tyDef of
-  G.TypeDefinitionScalar t -> TIScalar <$> fromScalarTyDef t loc
-  G.TypeDefinitionObject t -> return $ TIObj $ fromObjTyDef t loc
-  G.TypeDefinitionInterface t ->
-    return $ TIIFace $ fromIFaceDef t loc
-  G.TypeDefinitionUnion t -> return $ TIUnion $ fromUnionTyDef t
-  G.TypeDefinitionEnum t -> return $ TIEnum $ fromEnumTyDef t loc
-  G.TypeDefinitionInputObject t -> return $ TIInpObj $ fromInpObjTyDef t loc
+fromTyDef tyDef loc = pure $ case tyDef of
```
Comment (rakeshkky) [inline server/src-lib/Hasura/GraphQL/Validate/Types.hs]: Avoid `pure`.

File: server/src-lib/Hasura/RQL/DDL/Action.hs [hunk 14]
```diff
@@ -0,0 +1,302 @@
...
+--   , _rcIntervalSec :: !Word64
+--   , _rcTimeoutSec  :: !(Maybe Word64)
+--   } deriving (Show, Eq, Lift)
+
+-- data WebhookConf
+--   = WebhookConf
+--   { _wcUrl     :: !Text
+--   , _wcTimeout :: !Word64
+--   , _wcRetry   :: !RetryConf
+--   } deriving (Show, Eq)
```
Comment (rakeshkky) [inline server/src-lib/Hasura/RQL/DDL/Action.hs]: remove commented code

File: server/src-lib/Hasura/RQL/DDL/Action.hs [hunk 15]
```diff
@@ -0,0 +1,302 @@
...
+  getObjectTypeInfo responseBaseType
+  -- TODO: validate the output type
+  -- responseTypeInfo <- getNonObjectTypeInfo responseBaseType
+  -- case responseTypeInfo of
+  --   VT.TIScalar typeInfo -> return $ ActionOutputScalar typeInfo
+  --   VT.TIEnum typeInfo   -> return $ ActionOutputEnum typeInfo
+  --   VT.TIObj typeInfo    -> return $ ActionOutputObject typeInfo
+  --   _ -> throw400 InvalidParams $ "the output type: " <>
+  --        showNamedTy responseBaseType <>
+  --        " should be a scalar/enum/object"
```
Comment (rakeshkky) [inline server/src-lib/Hasura/RQL/DDL/Action.hs]: Try to understand this.

File: server/src-lib/Hasura/RQL/Types/Action.hs [hunk 16]
```diff
@@ -0,0 +1,214 @@
...
+data ActionInfo
+  = ActionInfo
+  { _aiName        :: !ActionName
+  , _aiDefinition  :: !ResolvedActionDefinition
+  , _aiPermissions :: !ActionPermissionMap
+  } deriving (Show, Eq)
+$(J.deriveToJSON (J.aesonDrop 3 J.snakeCase) ''ActionInfo)
+$(makeLenses ''ActionInfo)
+
+data InputWebhook
```
Comment (rakeshkky) [inline server/src-lib/Hasura/RQL/Types/Action.hs]: Use `URLTemplate` for input webhook.

File: server/src-lib/Hasura/RQL/Types/Action.hs [hunk 17]
```diff
@@ -0,0 +1,214 @@
...
+  } deriving (Show, Eq, Lift, Generic)
+instance NFData ActionPermissionDefinition
+instance Cacheable ActionPermissionDefinition
+$(J.deriveJSON (J.aesonDrop 4 J.snakeCase) ''ActionPermissionDefinition)
+
+data CreateActionPermission
+  = CreateActionPermission
+  { _capAction     :: !ActionName
+  , _capRole       :: !RoleName
+  , _capDefinition :: !ActionPermissionDefinition
```
Comment (rakeshkky) [inline server/src-lib/Hasura/RQL/Types/Action.hs]: Avoid `Definition`.

**Comment by SidhNor** [comment]
Thank your for great effort out there. This looks awesome.
When do you think this can realistically get into beta?

File: server/src-lib/Hasura/GraphQL/Context.hs [hunk 18]
```diff
@@ -57,6 +51,15 @@ instance ToJSON GCtx where
 
 type GCtxMap = Map.HashMap RoleName GCtx
 
+-- data GCtxMap
```
Comment (0x777) [inline server/src-lib/Hasura/GraphQL/Context.hs]: not needed

**Pull request merged**